File: others/past202005-open/past202005i.py
from sys import stdin
readline = stdin.readline

N = int(readline())
Q = int(readline())

result = []
transposed = False
rows = list(range(N + 1))
cols = list(range(N + 1))
for _ in range(Q):
    Query = list(map(int, readline().split()))
    if Query[0] == 1:
        A, B = Query[1:]
        if not transposed:
            t = rows[A]
            rows[A] = rows[B]
            rows[B] = t
        else:
            t = cols[A]
            cols[A] = cols[B]
            cols[B] = t
    elif Query[0] == 2:
        A, B = Query[1:]
        if not transposed:
            t = cols[A]
            cols[A] = cols[B]
            cols[B] = t
        else:
            t = rows[A]
            rows[A] = rows[B]
            rows[B] = t
    elif Query[0] == 3:
        transposed = not transposed
    elif Query[0] == 4:
        A, B = Query[1:]
        y, x = A, B
        if transposed:
            y, x = x, y
        y = rows[y]
        x = cols[x]
        result.append(N * (y - 1) + x - 1)
print('\n'.join(str(i) for i in result))

File: past/past202005/past202005i.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	defer flush()

	N := readInt()
	Q := readInt()

	transposed := false
	rows := make([]int, N+1)
	cols := make([]int, N+1)
	for i := 0; i < N+1; i++ {
		rows[i] = i
		cols[i] = i
	}

	for k := 0; k < Q; k++ {
		Q := readInt()
		if Q == 1 {
			A := readInt()
			B := readInt()
			if !transposed {
				t := rows[A]
				rows[A] = rows[B]
				rows[B] = t
			} else {
				t := cols[A]
				cols[A] = cols[B]
				cols[B] = t
			}
		} else if Q == 2 {
			A := readInt()
			B := readInt()
			if !transposed {
				t := cols[A]
				cols[A] = cols[B]
				cols[B] = t
			} else {
				t := rows[A]
				rows[A] = rows[B]
				rows[B] = t
			}
		} else if Q == 3 {
			transposed = !transposed
		} else if Q == 4 {
			A := readInt()
			B := readInt()
			i, j := A, B
			if transposed {
				i, j = j, i
			}
			i = rows[i]
			j = cols[j]
			println(N*(i-1) + j - 1)
		}
	}
}

const (
	ioBufferSize = 1 * 1024 * 1024 // 1 MB
)

var stdinScanner = func() *bufio.Scanner {
	result := bufio.NewScanner(os.Stdin)
	result.Buffer(make([]byte, ioBufferSize), ioBufferSize)
	result.Split(bufio.ScanWords)
	return result
}()

func readString() string {
	stdinScanner.Scan()
	return stdinScanner.Text()
}

func readInt() int {
	result, err := strconv.Atoi(readString())
	if err != nil {
		panic(err)
	}
	return result
}

var stdoutWriter = bufio.NewWriter(os.Stdout)

func flush() {
	stdoutWriter.Flush()
}

func println(args ...interface{}) (int, error) {
	return fmt.Fprintln(stdoutWriter, args...)
}
